core/trino-parser/src/main/java/io/trino/sql/SqlFormatter.java

```diff
@@ -227,6 +231,102 @@ protected Void visitLateral(Lateral node, Integer indent)
             return null;
         }
 
+        @Override
+        protected Void visitTableFunctionInvocation(TableFunctionInvocation node, Integer indent)
+        {
+            append(indent, "TABLE(");
+            appendTableFunctionInvocation(node, indent + 1);
+            builder.append(")");
+            return null;
+        }
+
+        private void appendTableFunctionInvocation(TableFunctionInvocation node, Integer indent)
+        {
+            builder.append(formatName(node.getName()))
+                    .append("(\n");
+            appendTableFunctionArguments(node.getArguments(), indent + 1);
+            if (!node.getCopartitioning().isEmpty()) {
+                builder.append("\n");
+                append(indent + 1, "COPARTITION ");
+                builder.append(node.getCopartitioning().stream()
+                        .map(tableList -> tableList.stream()
+                                .map(SqlFormatter::formatName)
+                                .collect(Collectors.joining(", ", "(", ")")))
+                        .collect(Collectors.joining(", ")));
+            }
+            builder.append(")");
+        }
+
+        private void appendTableFunctionArguments(List<Argument> arguments, int indent)
+        {
+            for (int i = 0; i < arguments.size(); i++) {
+                Argument argument = arguments.get(i);
+                if (argument.getName().isPresent()) {
+                    append(indent, ExpressionFormatter.formatExpression(argument.getName().get()));
+                    builder.append(" => ");
+                }
+                else {
+                    append(indent, "");
+                }
+                Node value = argument.getValue();
+                if (value instanceof TableArgument) {
```

Comment:
Ideally, the formatting of each argument should be done by a recursive call to format each element, instead of switching on the specific concrete classes here.

Reply:
Done for all except Expression arguments. I order to be processed recursively, they would require wrapping in another AST node, which I considered too much overhead.